File: src/assets/typescript/coordsLoader.ts
import { promises as fs } from "node:fs";
import path from "node:path";

export type CoordinatePack = {
  manifest: any;
  zoneSpaces: any[];
  projectionBounds: any[];
  mapDefaults: any[];
  legacyBases: any[];
  instanceAnchors: any[];
  zoneSpaceByAreaId: Map<number, any>;
  projectionBoundsByKey: Map<string, any>;
  mapDefaultByMapId: Map<number, number>;
  legacyBasisByKey: Map<number, any>;
  instanceAnchorByMapId: Map<number, any>;
  instanceAnchorByZoneAreaId: Map<number, any>;
};

const FILES: Record<string, string> = {
  manifest: "manifest.json",
  zoneSpaces: "zone_spaces.json",
  projectionBounds: "projection_bounds.json",
  mapDefaults: "map_defaults.json",
  legacyBases: "legacy_bases.json",
  instanceAnchors: "instance_anchors.json",
};
const CURRENT_SCHEMA_VERSION = 2;

export async function loadCoordinatePack(packDir: string): Promise<CoordinatePack> {
  const pack: any = {};
  for (const [key, filename] of Object.entries(FILES)) {
    const filePath = path.join(packDir, filename);
    pack[key] = JSON.parse(await fs.readFile(filePath, "utf8"));
  }
  const schemaVersion = Number(pack.manifest.schemaVersion);
  if (schemaVersion !== CURRENT_SCHEMA_VERSION) {
    throw new Error(
      `Unsupported coordinate pack schemaVersion=${schemaVersion} in ${path.join(packDir, FILES.manifest)}; expected ${CURRENT_SCHEMA_VERSION}`,
    );
  }

  pack.zoneSpaceByAreaId = new Map<number, any>(
    pack.zoneSpaces.map((row: any) => [Number(row.zoneAreaId), row]),
  );
  pack.projectionBoundsByKey = new Map<string, any>(
    pack.projectionBounds.map((row: any) => [
      `${Number(row.mapId)}:${Number(row.uiMapId)}`,
      row,
    ]),
  );
  pack.mapDefaultByMapId = new Map<number, number>(
    pack.mapDefaults.map((row: any) => [Number(row.mapId), Number(row.coordUiMapId)]),
  );
  pack.legacyBasisByKey = new Map<number, any>(
    pack.legacyBases.map((row: any) => [Number(row.legacyKey), row]),
  );
  pack.instanceAnchorByMapId = new Map<number, any>(
    pack.instanceAnchors.map((row: any) => [Number(row.instanceMapId), row]),
  );
  pack.instanceAnchorByZoneAreaId = new Map<number, any>(
    pack.instanceAnchors
      .filter((row: any) => row.zoneAreaId !== null && row.zoneAreaId !== undefined)
      .map((row: any) => [Number(row.zoneAreaId), row]),
  );
  return pack as CoordinatePack;
}

File: src/assets/python/coords_loader.py
from __future__ import annotations

import json
from pathlib import Path
from typing import Any


FILES = {
    "manifest": "manifest.json",
    "zoneSpaces": "zone_spaces.json",
    "projectionBounds": "projection_bounds.json",
    "mapDefaults": "map_defaults.json",
    "legacyBases": "legacy_bases.json",
    "instanceAnchors": "instance_anchors.json",
}
CURRENT_SCHEMA_VERSION = 2


def load_coordinate_pack(pack_dir: str | Path) -> dict[str, Any]:
    root = Path(pack_dir)
    pack = {
        key: _read_json(root / filename)
        for key, filename in FILES.items()
    }
    schema_version = int(pack["manifest"]["schemaVersion"])
    if schema_version != CURRENT_SCHEMA_VERSION:
        raise ValueError(
            f"Unsupported coordinate pack schemaVersion={schema_version} in "
            f"{root / FILES['manifest']}; expected {CURRENT_SCHEMA_VERSION}"
        )

    pack["zoneSpaceByAreaId"] = {
        int(row["zoneAreaId"]): row for row in pack["zoneSpaces"]
    }
    pack["projectionBoundsByKey"] = {
        (int(row["mapId"]), int(row["uiMapId"])): row
        for row in pack["projectionBounds"]
    }
    pack["mapDefaultByMapId"] = {
        int(row["mapId"]): int(row["coordUiMapId"])
        for row in pack["mapDefaults"]
    }
    pack["legacyBasisByKey"] = {
        int(row["legacyKey"]): row for row in pack["legacyBases"]
    }
    pack["instanceAnchorByMapId"] = {
        int(row["instanceMapId"]): row for row in pack["instanceAnchors"]
    }
    pack["instanceAnchorByZoneAreaId"] = {
        int(row["zoneAreaId"]): row
        for row in pack["instanceAnchors"]
        if row.get("zoneAreaId") is not None
    }
    return pack


def _read_json(path: Path) -> Any:
    with path.open("r", encoding="utf-8") as handle:
        return json.load(handle)
